selfdrive/car/hyundai/values.py

```diff
@@ -348,35 +344,66 @@ class Buttons:
+                                          DATE_FW_PATTERN.search(fw))
     if code_match is not None:
-      code = code_match.group()
+      code: bytes = code_match.group()
+      part = part_match.group() if part_match else None
       date = date_match.group() if date_match else None
-      codes[code].add(date)
+      if part is not None:
+        # part number starts with generic ECU part type, add what is specific to platform
+        code += b"-" + part[-5:]
 
-  # Create platform codes for all dates inclusive if ECU has FW dates
-  final_codes = set()
-  for code, dates in codes.items():
-    # Radar and some cameras don't have FW dates
-    if None in dates:
-      final_codes.add(code)
-      continue
+      codes.add((code, date))
+  return codes
 
-    try:
-      parsed = {datetime.strptime(cast(bytes, date).decode()[:4], '%y%m') for date in dates}
-    except ValueError:
-      cloudlog.exception(f'Error parsing date in FW versions: {code!r}, {dates}')
-      final_codes.add(code)
-      continue
 
-    for monthly in rrule.rrule(rrule.MONTHLY, dtstart=min(parsed), until=max(parsed)):
-      final_codes.add(code + b'-' + monthly.strftime('%y%m').encode())
+def match_fw_to_car_fuzzy(fw_versions_dict, log=True) -> Set[str]:
+  # Non-electric CAN FD platforms often do not have platform code specifiers needed
+  # to distinguish between hybrid and ICE. All EVs so far are either exclusively
+  # electric or specify electric in the platform code.
+  invalid = set(CANFD_CAR - EV_CAR)
+  candidates = FW_VERSIONS
 
-  return final_codes
+  for candidate, fws in candidates.items():
+    for ecu, expected_versions in fws.items():
+      addr = ecu[1:]
+      # Only check ECUs expected to have platform codes
+      if ecu[0] not in PLATFORM_CODE_ECUS:
+        continue
+
+      # Expected platform codes & dates
+      codes = get_platform_codes(expected_versions)
+      expected_platform_codes = {code for code, _ in codes}
+      expected_dates = {date for _, date in codes if date is not None}
+
+      # Found platform codes & dates
+      codes = get_platform_codes(fw_versions_dict.get(addr, set()))
+      found_platform_codes = {code for code, _ in codes}
+      found_dates = {date for _, date in codes if date is not None}
+
+      # Check platform code + part number matches for any found versions
+      if not any(found_platform_code in expected_platform_codes for found_platform_code in found_platform_codes):
+        invalid.add(candidate)
+        break
+
+      # Don't check dates if none in database
+      if len(expected_dates):
+        if not len(found_dates):
+          invalid.add(candidate)
+          break
+
+        # Check all dates within range in the database, format is %y%m%d
+        if not all(min(expected_dates) <= found_date <= max(expected_dates) for found_date in found_dates):
+          invalid.add(candidate)
+          break
+
+  return set(candidates.keys()) - invalid
```

Comment:
can you add some kind of whitelist for matches? I see that it all happens to work out now, but seems easy to write an innocent bug that will introduce FPs

Reply:
that's really bad

Reply:
This is what exact matching does (and that candidates indirection), I'll fix that too

Reply:
Think I did this. For each candidate, it keeps track of ECUs with platform codes in the database, and asserts that no breaks are reached while checking all the live versions for that ECU (by adding those ECUs to a set at the end).

A 'whitelist' for the CAN FD/hybrid stuff would look pretty similar `matches & (set(DBC) - set(CANFD_CAR - EV_CAR))`